bwt/wavelet.go

```diff
@@ -0,0 +1,340 @@
+package bwt
+
+import (
+	"fmt"
+	"math"
+
+	"golang.org/x/exp/slices"
+)
+
+// waveletTree datastructure that allows us to
+// conduct RSA queries on strings.
+//
+// For the waveletTree's usage, please read the its
+// method documentation. To understand what it is and how
+// it works, then read below.
+//
+// # The WaveletTree has several imporant components
+//
+// ## The Character's Path Encoding
+//
+// one important component is a character's path encoding.
+// which character we are woring with in a given path in the tree.
+// For example, given the alphabet A B C D E F G, a possible encoding is:
+// A: 000
+// B: 001
+// C: 010
+// D: 011
+// E: 100
+// F: 101
+// G: 110
+// H: 111
+//
+// If we wanted to get to the leaf that represent the character D, we'd
+// take the path:
+//
+//	   root
+//	  /
+//	left
+//	  \
+//	 right
+//	    \
+//	   right
+//
+// ## The Data Represented at each node
+//
+// Let us consider the sequence "bananas"
+// It has the alphabet b, a, n, s
+// Let's say it has the encoding:
+// a: 00
+// n: 01
+// b: 10
+// s: 11
+// and that 0 if left and 1 is right
+// We can represent this tree with bitvectors:
+//
+//	     0010101
+//	    /       \
+//	  1000      001
+//	 /    \    /   \
+//	a      n  b     s
+//
+// If we translate each bit vector to its corresponding string, then it becomes:
+//
+//	     bananas
+//	    /       \
+//	  baaa      nns
+//	 /    \    /   \
+//	a      b  n     s
+//
+// ## RSA Intuition
+//
+// From here you may be able to build some intuition as to how we can take RSA queries given
+// a characters path encoding and which character we'd like to Rank, Select, and Access.
```

Comment:
RSA is spelled out here! I was beginning to think I wouldn't find it.

Reply:
Sorry about that. Yeah, I was thinking that was more appropriate to spell out in the RSABitvector and just telling readers to look there to learn more about RSA. Happy to make that redundant though so maintainers don't have to jump around to Grok all of these concepts.

Reply:
I put Definitions of Acronyms towards the top of explanations like this. LMK if that works out.